packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,462 @@
+  private logger: Logger
+  private didRepository: DidRepository
+
+  public constructor(
+    @inject(InjectionSymbols.Wallet) wallet: Wallet,
+    config: AgentConfig,
+    connectionService: ConnectionService,
+    didRepository: DidRepository,
+    jwsService: JwsService
+  ) {
+    this.config = config
+    this.connectionService = connectionService
+    this.didRepository = didRepository
+    this.jwsService = jwsService
+    this.logger = config.logger
+  }
+
+  public async createRequest(
+    connectionRecord: ConnectionRecord,
+    params: DidExchangeRequestParams
+  ): Promise<DidExchangeRequestMessage> {
+    this.logger.debug(`Create message ${DidExchangeRequestMessage.type} start`, connectionRecord)
+
+    if (!connectionRecord.invitation) {
+      throw new AriesFrameworkError('Connection invitation is missing.')
+    }
+
+    DidExchangeStateMachine.assertCreateMessageState(DidExchangeRequestMessage.type, connectionRecord)
+
+    const { goal, goalCode, routing, autoAcceptConnection } = params
+    const label = params.label ?? this.config.label
+    const { verkey } = routing
+    const peerDid = await this.createPeerDidDoc(routing)
+    const parentThreadId = connectionRecord.invitation?.id
+
+    const message = new DidExchangeRequestMessage({ label, parentThreadId, did: peerDid.did, goal, goalCode })
+
+    // Create sign attachment containing didDoc
+    if (peerDid.numAlgo === PeerDidNumAlgo.GenesisDoc) {
+      const didDocAttach = await this.createSignedAttachment(peerDid.didDocument, verkey)
+      message.didDoc = didDocAttach
+    }
+
+    connectionRecord.did = peerDid.did
+    connectionRecord.verkey = verkey
+    connectionRecord.threadId = message.id
+
+    if (autoAcceptConnection !== undefined || autoAcceptConnection !== null) {
+      connectionRecord.autoAcceptConnection = autoAcceptConnection
+    }
+
+    await this.updateState(DidExchangeRequestMessage.type, connectionRecord)
+    this.logger.debug(`Create message ${DidExchangeRequestMessage.type} end`, {
+      connectionRecord,
+      message,
+    })
+    return message
+  }
+
+  public async processRequest(
+    messageContext: InboundMessageContext<DidExchangeRequestMessage>,
+    routing?: Routing
+  ): Promise<ConnectionRecord> {
+    this.logger.debug(`Process message ${DidExchangeRequestMessage.type} start`, messageContext)
+
+    // eslint-disable-next-line prefer-const
+    let { connection: connectionRecord, message } = messageContext
+
+    if (!connectionRecord) {
+      throw new AriesFrameworkError('No connection record in message context.')
+    }
+
+    DidExchangeStateMachine.assertProcessMessageState(DidExchangeRequestMessage.type, connectionRecord)
+
+    // check corresponding invitation ID is the request's ~thread.pthid
+
+    if (connectionRecord.invitation?.id !== message.thread?.parentThreadId) {
+      throw new ProblemReportError('Missing reference to invitation.', { problemCode: 'request_not_accepted' })
```

Comment:
Throwing a `ProblemReportError` will send a generic problem report message. According to the DIDExchange RFC we should send an adopted DID Exchange or OOB problem report:

> The problem-report may be an adopted Out of Band protocol message or an adopted DID Exchange protocol message, depending on where in the processing of the invitation the error was detected.

See e.g. `ConnectionProblemReportError.ts` for an example of an adopted problem report error usage

Reply:
👍  Updated.